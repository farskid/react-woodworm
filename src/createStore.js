/* @flow */
import React, { createContext, Component } from 'react'

import validateStore from './validateStore'

type InterfaceShape = {
  Consumer: any,
  Provider: any,
}

type OptionsType = {
  model?: any,
  actions?: Object,
}

export default function createStore(
  options: OptionsShape = {}
): InterfaceShape {
  const Store = createContext()

  const actions = options.actions || {}

  return {
    Provider: class Provider extends Component {
      state: {
        state: any,
        actions: Object,
      }

      constructor(props, context) {
        super(props, context)

        const initialState = props.initialState || options.model
        const resolvedActions = Object.keys(actions).reduce((map, name) => {
          map[name] = payload => {
            const newState = actions[name](this.state.state, ...payload)
            const isAsyncAction = newState.then !== undefined
            if (isAsyncAction) {
              newState.then(resolvedState => {
                this.setState({ state: resolvedState })
              })
            } else {
              this.setState({ state: newState })
            }
          }

          return map
        }, {})

        this.state = {
          state: initialState,
          actions: resolvedActions,
        }
      }

      render() {
        return (
          <Store.Provider value={this.state}>
            {this.props.children}
          </Store.Provider>
        )
      }
    },

    Consumer: ({ children }) => (
      <Store.Consumer>
        {store => validateStore(store) && children(store.state, store.actions)}
      </Store.Consumer>
    ),
  }
}
