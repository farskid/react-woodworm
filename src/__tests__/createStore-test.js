import React from 'react'
import TestRenderer from 'react-test-renderer'
import { Simulate } from 'react-dom/test-utils'

import createStore from '../createStore'

describe('Creating a store', () => {
  it('should return a Consumer and a Provider', () => {
    const store = createStore()

    expect(Object.keys(store)).toEqual(['Provider', 'Consumer'])
  })
})

describe('Rendering the <Consumer>', () => {
  it('should throw without the <Provider>', () => {
    const { Provider, Consumer } = createStore()

    const render = () =>
      TestRenderer.create(<Consumer>{store => null}</Consumer>)

    expect(render).toThrow()
  })

  it('should render the initial state', () => {
    const { Provider, Consumer } = createStore({
      foo: 1,
      bar: 'bar',
      baz: {
        foobar: 5,
      },
    })

    const tree = TestRenderer.create(
      <Provider>
        <Consumer>
          {state => <div>{JSON.stringify(state, null, 2)}</div>}
        </Consumer>
      </Provider>
    )

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('should render the updated state', () => {
    const { Provider, Consumer } = createStore(
      {
        foo: 1,
        bar: 'bar',
        baz: {
          foobar: 5,
        },
      },
      {
        update: (state, payload) => Object.assign({}, state, { foo: payload }),
      }
    )

    const tree = TestRenderer.create(
      <Provider>
        <Consumer>
          {(state, actions) => (
            <div>
              <div>{JSON.stringify(state, null, 2)}</div>
              <button onClick={() => actions.update(10)}>Update</button>
            </div>
          )}
        </Consumer>
      </Provider>
    )

    tree.root.findByType('button').props.onClick()

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('should create a separate store for every Provider instance', () => {
    const { Provider, Consumer } = createStore(0, {
      increment: state => state + 1,
    })

    const Counter = () => (
      <Provider>
        <Consumer>
          {(state, actions) => (
            <div>
              <div>Count: {state}</div>
              <button onClick={actions.increment}>+</button>
            </div>
          )}
        </Consumer>
      </Provider>
    )
    const tree = TestRenderer.create(
      <div>
        <Counter />
        <Counter />
      </div>
    )

    tree.root.findAllByType('button')[0].props.onClick()

    expect(tree.toJSON()).toMatchSnapshot()
  })
})
