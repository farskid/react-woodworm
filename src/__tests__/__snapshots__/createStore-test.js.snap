// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the <Consumer> should create a separate store for every Provider instance 1`] = `
<div>
  <div>
    <div>
      Count: 
      1
    </div>
    <button
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div>
    <div>
      Count: 
      0
    </div>
    <button
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;

exports[`Rendering the <Consumer> should render the initial state 1`] = `
<div>
  {
  "foo": 1,
  "bar": "bar",
  "baz": {
    "foobar": 5
  }
}
</div>
`;

exports[`Rendering the <Consumer> should render the initial state passed to the Provider 1`] = `
<div>
  {
  "foo": 5
}
</div>
`;

exports[`Rendering the <Consumer> should render the updated state 1`] = `
<div>
  <div>
    {
  "foo": 10,
  "bar": "bar",
  "baz": {
    "foobar": 5
  }
}
  </div>
  <button
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;
